drivers/net/dsa/lantiq_gsw_mdio.c

```diff
@@ -35,19 +36,103 @@
 #include "lantiq_gsw.h"
 #include "lantiq_pce.h"
 
-static u32 gsw_mdio_read(struct gswip_priv *priv, void *address)
+#define NUM_ACCESSIBLE_REGS (30)
+#define TARGET_BASE_ADDRESS_REG (31)
+
+struct gsw_mdio {
+	struct mdio_device *mdio_dev;
+	struct gswip_priv common;
+};
+
+static u32 gsw_mdio_read_actual(struct mdio_device *mdio, u32 reg)
 {
-	/* TODO WARP-5829: write this function */
-	return 0;
+	return mdio->bus->read(mdio->bus, mdio->addr, reg);
 }
 
-static void gsw_mdio_write(struct gswip_priv *priv, u32 val, void *address)
+static void gsw_mdio_write_actual(struct mdio_device *mdio, u32 reg, u32 val)
 {
-	/* TODO WARP-5829: write this function */
+	mdio->bus->write(mdio->bus, mdio->addr, reg, val);
+}
+
+static inline u32 gsw_mdio_read_tbar(struct mdio_device *mdio)
+{
+	return gsw_mdio_read_actual(mdio, TARGET_BASE_ADDRESS_REG);
+}
+
+static inline void gsw_mdio_write_tbar(struct mdio_device *mdio, u32 reg_addr)
+{
+	gsw_mdio_write_actual(mdio, TARGET_BASE_ADDRESS_REG, reg_addr);
+}
+
+static u32 gsw_mdio_check_write_tbar(struct mdio_device *mdio, u32 reg_addr)
+{
+	u32 tbar = gsw_mdio_read_tbar(mdio);
```

Comment:
TBAR is always read from the part (vs. keeping a soft copy in the driver state data, even if global) to protect against the part resetting or otherwise going off the rails and us performing a read/write from an incorrect register.

Reply:
Also makes the driver more compatible with DSA clustering if someone wanted to add that functionality